File: rotate-list.ts
function rotateRight(head: ListNode | null, k: number): ListNode | null {
    if(head === null) return null;
    let root = head;
    let len = 1;
    while(root.next !== null) {
        root = root.next;
        len++;
    }
    let requiredRotations = len - (k % len);
    if(requiredRotations == 0) {
        return head;
    }
    root.next = head;
    while(requiredRotations) {
        root = root.next;
        --requiredRotations;
    }
    head = root.next;
    root.next = null;
    return head;
};

File: rotate-list.rs
pub fn rotate_right(mut h: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
    if h.is_none() || k == 0 {
        return h;
    };
    let (mut l, mut x) = (0, &h);
    while let Some(n) = x {
        x = &n.next;
        l += 1
    }
    if l < 2 || k % l == 0 {
        return h;
    };
    let mut x = &mut h;
    for _ in 0..l - (k % l) {
        if let Some(n) = x {
            x = &mut n.next
        }
    }
    let mut res = x.take();
    let mut x = &mut res;
    while let Some(n) = x {
        x = &mut n.next
    }
    *x = h;
    res
}
